### PR TITLE
K-borda method for choosing committee

### PR DESCRIPTION
**Pull request by msluszniak**

### CODE REVIEW

File: votepy/k-borda.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
+from typing import Iterable, Sequence, List
+from ordinal_election import OrdinalElection, OrdinalBallot
+
+
+def k_borda(voting: OrdinalElection, size_of_committee: int, number_of_scored_candidates: int) -> OrdinalBallot:
```
Comment (Wolokin) [inline votepy/k-borda.py]: The user shouldn't be concerned with our internal classes. We shouldn't expect the user to supply us with the OrdinalElection - he might, but most likely won't

File: votepy/k-borda.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        raise ValueError(f"Size of committee needs to be from the range 1 to the number of all candidates.")
+    if number_of_scored_candidates > n or number_of_scored_candidates <= 0:
+        raise ValueError(f"Number of candidates scored in k-borda needs to be from range 1 to the number of all candidates.")
+    for vote in voting:
+        m = number_of_scored_candidates
+        for candidate in vote[:number_of_scored_candidates]:
+            results[candidate] += m
+            m-=1
+    _, committee =  zip(*sorted([(val, idx) for idx, val in enumerate(results)], reverse=True))
+    return OrdinalBallot(committee[:size_of_committee])
```
Comment (Wolokin) [inline votepy/k-borda.py]: This will fail most of the time due to checks added in OrdinalBallot. You might wanna rebase onto dev.

Comment (Wolokin) [reply votepy/k-borda.py]: The added check requires that the ballot is a permutation of [0, length-1]

File: votepy/k-borda.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
+from typing import Iterable, Sequence, List
```
Comment (Wolokin) [inline votepy/k-borda.py]: Imports from 'typing' are deprecated. See https://docs.python.org/3/library/typing.html#typing.Iterable

File: votepy/ordinal_election.py [hunk 4]
```diff
@@ -33,6 +33,8 @@ def __init__(self, preference_orders: list, mapping=None):
                 if candidate not in self.candidates:
                     raise ValueError(f"Candidate sets in preference orders differ: candidate {candidate} is not present in every preference order")
             self.append(OrdinalBallot(preference, mapping))
+        self.ballot_size = len(preference_orders[0])
+        self.number_of_votes = len(preference_orders)
```
Comment (Wolokin) [inline votepy/ordinal_election.py]: number_of_voters might be a better name, as number_of_votes may be confused with the number of candidates a single voter had to vote for

**Review by Wolokin — CHANGES_REQUESTED** [review]

File: votepy/k-borda.py (hunk 2 above)
Comment (msluszniak) [reply votepy/k-borda.py]: Should I return simply a list representing the chosen committee?

File: votepy/k-borda.py [hunk 5]
```diff
@@ -1,8 +1,7 @@
-from typing import Iterable, Sequence, List
 from ordinal_election import OrdinalElection, OrdinalBallot
+from typing import Union, List
```
Comment (Wolokin) [inline votepy/k-borda.py]: List is deprecated as well for the same reason as before - you can just use the built-in list.

File: votepy/k-borda.py [hunk 6]
```diff
@@ -1,8 +1,7 @@
...
 
-
-def k_borda(voting: OrdinalElection, size_of_committee: int, number_of_scored_candidates: int) -> OrdinalBallot:
+def k_borda(voting: Union[OrdinalElection, List[int]], size_of_committee: int, number_of_scored_candidates: int) -> OrdinalBallot:
```
Comment (Wolokin) [inline votepy/k-borda.py]: You can use OrdinalElection | List[int]. I belive it's cleaner. See: https://docs.python.org/3/library/stdtypes.html#types-union

**Review by Wolokin — COMMENTED** [review]
I can merge this, but please have a look at my suggestions beforehand

File: votepy/k-borda.py (hunk 6 above)
Comment (msluszniak) [reply votepy/k-borda.py]: Seems good

File: votepy/k-borda.py (hunk 5 above)
Comment (msluszniak) [reply votepy/k-borda.py]: Oh, I miss this one. You're right.

**Comment by Janix4000** [comment]
[Implement k-Bord rule](https://trello.com/c/7qlBXkHc/5-implement-k-bord-rule)

File: votepy/k-borda.py (hunk 6 above)
Comment (Janix4000) [reply votepy/k-borda.py]: So we need to update the compatible Python versions - `|` syntax is available from Python 3.9.

File: votepy/k-borda.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    if size_of_committee > n or size_of_committee <= 0:
+        raise ValueError(f"Size of committee needs to be from the range 1 to the number of all candidates.")
+    if number_of_scored_candidates > n or number_of_scored_candidates <= 0:
+        raise ValueError(f"Number of candidates scored in k-borda needs to be from range 1 to the number of all candidates.")
+    for vote in voting:
+        m = number_of_scored_candidates
+        for candidate in vote[:number_of_scored_candidates]:
+            results[candidate] += m
+            m-=1
+    _, committee =  zip(*sorted([(val, idx) for idx, val in enumerate(results)], reverse=True))
```
Comment (Janix4000) [inline votepy/k-borda.py]: `committee, _ = zip(*sorted(enumerate(results), reverse=True, key=lambda t: t[1]))` would be more readable, I believe

**Review by Janix4000 — CHANGES_REQUESTED** [review]

File: votepy/k-borda.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
+from ordinal_election import OrdinalElection, OrdinalBallot
+from typing import Union, List
+
+def k_borda(voting: Union[OrdinalElection, List[int]], size_of_committee: int, number_of_scored_candidates: int) -> List[int]:
```
Comment (Janix4000) [inline votepy/k-borda.py]: Change `List` -> to `list`

**Pull request merged**